subsys/bluetooth/services/ras/rrsp/ras_rrsp.c

```diff
@@ -8,21 +8,674 @@
+		net_buf_simple_add(&segment_buf, sizeof(struct ras_segment) + max_data_len);
+	if (!ras_segment) {
+		LOG_ERR("Cannot allocate segment buffer");
+		return -ENOMEM;
+	}
+
+	bool first_seg = (rrsp->active_buf_read_cursor == 0);
+	bool last_seg;
+	uint16_t actual_data_len =
+		bt_ras_rd_buffer_bytes_pull(rrsp->active_buf, ras_segment->data, max_data_len,
+					    &rrsp->active_buf_read_cursor, &last_seg);
+
+	LOG_DBG("Got %u bytes (max: %u)", actual_data_len, max_data_len);
+
+	if (actual_data_len) {
+		ras_segment->header.first_seg = first_seg;
+		ras_segment->header.last_seg = last_seg;
+		ras_segment->header.seg_counter = rrsp->segment_counter & BIT_MASK(6);
+
+		(void)net_buf_simple_remove_mem(&segment_buf, (max_data_len - actual_data_len));
+
+		err = ondemand_rd_notify_or_indicate(rrsp->conn, &segment_buf);
+		if (err) {
+			LOG_WRN("ondemand_rd_notify_or_indicate failed err %d", err);
+
+			/* Keep retrying */
+			rrsp->active_buf_read_cursor -= actual_data_len;
+
+			return err;
+		}
+
+		rrsp->segment_counter++;
+
+		LOG_DBG("Segment with RSC %d sent", rrsp->segment_counter);
+	}
+
+	if (!last_seg) {
+		k_work_submit_to_queue(&rrsp_wq, &rrsp->send_data_work);
+	} else {
+		LOG_DBG("All segments sent");
+		rascp_send_complete_rd_rsp(rrsp->conn, rrsp->active_buf->ranging_counter);
+		rrsp->streaming = false;
+		k_work_cancel(&rrsp->send_data_work);
+		k_timer_start(&rrsp->rascp_timeout, RASCP_ACK_DATA_TIMEOUT, K_NO_WAIT);
+	}
+
+	return 0;
+}
+
+static void send_data_work_handler(struct k_work *work)
+{
+	struct bt_ras_rrsp *rrsp = CONTAINER_OF(work, struct bt_ras_rrsp, send_data_work);
+
+	if (!rrsp->streaming || !rrsp->active_buf) {
+		return;
+	}
+
+	int err = rd_segment_send(rrsp);
+
+	if (err) {
+		/* Will keep retrying. */
+		LOG_WRN("Failed to send segment: %d", err);
+	}
+}
+
+static void rascp_work_handler(struct k_work *work)
+{
+	struct bt_ras_rrsp *rrsp = CONTAINER_OF(work, struct bt_ras_rrsp, rascp_work);
+
+	LOG_DBG("rrsp %p", rrsp);
+
+	rascp_cmd_handle(rrsp);
+}
+
+static void status_work_handler(struct k_work *work)
+{
+	struct bt_ras_rrsp *rrsp = CONTAINER_OF(work, struct bt_ras_rrsp, status_work);
+
+	LOG_DBG("rrsp %p", rrsp);
+
+	if (rrsp->notify_overwritten) {
+		int err = rd_status_notify_or_indicate(rrsp->conn, BT_UUID_RAS_RD_OVERWRITTEN,
+						       rrsp->overwritten_ranging_counter);
+		if (err) {
+			LOG_WRN("Notify overwritten failed: %d", err);
+		}
+
+		rrsp->notify_overwritten = false;
+	}
+
+	if (rrsp->notify_ready) {
+		int err = rd_status_notify_or_indicate(rrsp->conn, BT_UUID_RAS_RD_READY,
+						       rrsp->ready_ranging_counter);
+		if (err) {
+			LOG_WRN("Notify ready failed: %d", err);
+		}
+
+		rrsp->notify_ready = false;
+	}
+}
+
+static void rascp_timeout_handler(struct k_timer *timer)
+{
+	struct bt_ras_rrsp *rrsp = CONTAINER_OF(timer, struct bt_ras_rrsp, rascp_timeout);
+
+	LOG_DBG("rrsp %p", rrsp);
+
+	LOG_WRN("RAS-CP timeout");
+
+	bt_ras_rd_buffer_release(rrsp->active_buf);
```

Comment:
so if I understand correctly, this function is called from the system workqueue, right? So it can preempt RRSP WQ ?

Just checking but can you explain why it's safe to write to active_buf and active_buf_read_cursor below ?

Reply:
I think it was fine in the scenarios it could occur, but moving it to the RRSP WQ just in case, and to make it easier to reason about.